lib/Common/Memory/WriteBarrierMacros.h

```diff
@@ -17,6 +17,8 @@
 
 #define FieldWithBarrier(type, ...) \
     typename WriteBarrierFieldTypeTraits<type, ##__VA_ARGS__>::Type
+#define FieldWithBarrierT(type, ...) \
```

Comment:
That will not work because there are some declarations with FieldWithBarrier inside a template and some not within a template. For declarations within a template we require 'typename' whereas for those not inside a template 'typename' should not be used with legacy compiler.